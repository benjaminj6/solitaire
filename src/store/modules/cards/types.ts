export type CardRank =
  | 'A'
  | '2'
  | '3'
  | '4'
  | '5'
  | '6'
  | '7'
  | '8'
  | '9'
  | '10'
  | 'J'
  | 'Q'
  | 'K';

export type CardRankNumber =
  | '0'
  | '1'
  | '2'
  | '3'
  | '4'
  | '5'
  | '6'
  | '7'
  | '8'
  | '9'
  | '10'
  | '11'
  | '12';

export type Suit = 'hearts' | 'spades' | 'diamonds' | 'clubs';

// TODO - make the typing on this more strict to basically just be a suit and a rank separated by a dash
export type CardId = string;

export type Card = {
  id: CardId;
  suit: Suit;
  rank: CardRank;
  value: number;
  visible: boolean;
};

export type Deck = {
  [x: string]: Card;
};

export type CardRefArray = CardId[];

export type TableauRow = '1' | '2' | '3' | '4' | '5' | '6' | '7';
export type Tableau = {
  [row in TableauRow]: CardRefArray;
};

export type Stock = CardRefArray;
export type Hand = CardRefArray;

export type Foundation = { [suit in Suit]: CardRefArray };

export type State = {
  deck: Deck;
  dealt: boolean,
  tableau: Tableau;
  stock: Stock;
  hand: Hand;
  foundation: Foundation;
};
