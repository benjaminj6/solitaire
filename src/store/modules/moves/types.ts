import { CardId, CardLocation } from '../cards/types';

export type Move = {
  cards: CardId[];
  currentLocation?: CardLocation;
  destination?: CardLocation;
};

export type MoveId = string;

export type State = {
  nextMove: Move | {};
  moves: {
    [id: string]: Move;
  };
  history: Move[];
};
