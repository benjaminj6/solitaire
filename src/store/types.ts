// Generic Redux-related types
import { Store, Dispatch } from 'redux';

export type Action = {
  type: string;
  result?: {
    [x: string]: any;
  };
};

export type ActionCreator = (...args: any[]) => Action;

type GetState = Store<any>['getState'];
export type ActionThunk = (dispatch: Dispatch<Action>, getState: GetState) => Action;

export type ActionThunkCreator = (...args: any[]) => ActionThunk;

export type Reducer < T > = (state: T, action: Action) => T;
