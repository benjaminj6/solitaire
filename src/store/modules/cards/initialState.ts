import * as range from 'lodash/fp/range';
import { Card, CardRank, CardRankNumber, Deck, State, Suit } from './types';

type GetRankFromRemainder = (x: CardRankNumber) => CardRank;
export const getRankFromRemainder: GetRankFromRemainder = num => {
  switch (num) {
    case '0':
      return 'K';
    case '1':
      return 'Q';
    case '2':
      return 'J';
    case '12':
      return 'A';
    default:
      return `${13 - +num}` as CardRank;
  }
};

type GetInitialSuit = (x: Suit) => Deck;
export const getInitialSuit: GetInitialSuit = suitName => {
  return range(1, 14)
    .map((num: number) => {
      const value = `${num % 13}` as CardRankNumber;
      const rank = getRankFromRemainder(value);

      return {
        id: `${suitName}-${rank}`,
        rank,
        suit: suitName,
        value: 13 - +value,
        visible: false,
      };
    })
    .sort((first: Card, second: Card) => first.value - second.value)
    .reduce((obj: Deck, card: Card) => {
      obj[card.id] = card;
      return obj;
    }, {});
};

type GetInitialDeck = () => Deck;
export const getInitialDeck: GetInitialDeck = () => {
  const suits = ['hearts', 'spades', 'diamonds', 'clubs'];
  return suits.reduce(
    (obj: Deck, suitName: Suit) => ({
      ...obj,
      ...getInitialSuit(suitName),
    }),
    {}
  );
};

export default {
  deck: getInitialDeck(),
  shuffledDeck: [],
} as State;
