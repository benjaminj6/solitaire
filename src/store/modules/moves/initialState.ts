import { State } from './types';

const initialState: State = {
  nextMove: {},
  moves: {},
  history: [],
};

export default initialState;