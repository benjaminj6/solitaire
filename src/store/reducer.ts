import { combineReducers } from 'redux';

import cards, { State as CardState } from './modules/cards';
import moves, { State as MovesState } from './modules/moves';

export type State = {
  cards: CardState;
  moves: MovesState
};

export default combineReducers({ cards, moves });
