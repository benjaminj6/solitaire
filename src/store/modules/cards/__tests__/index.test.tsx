import { pick, cloneDeep } from 'lodash';
import initialState from '../initialState';
import { State } from '../types';
import reducer, { DEALCARDS, DEALCARDS_INVALID, DEAL_HAND } from '../index';

describe('reducer', () => {
  // The DEALCARDS relies on randomness to shuffle the deck
  // This means we have to make looser assertions rather than using snapshots
  describe('DEALERCARDS', () => {
    test('should return tableau rows that are filled', () => {
      const { tableau } = reducer(undefined, { type: DEALCARDS });
      const entries = Object.entries(tableau);

      entries.forEach(([key, value]) => {
        expect(value.length).toEqual(+key);
      });
    });

    test('should return a `stock` with 24 cards', () => {
      const { stock } = reducer(undefined, { type: DEALCARDS });

      expect(stock.length).toEqual(24);
    });

    test('should enable the `dealt` flag', () => {
      const { dealt } = reducer(undefined, { type: DEALCARDS });

      expect(dealt).toEqual(true);
    });
  });

  describe('DEALERCARDS_INVALID', () => {
    test('should return the state', () => {
      const result = reducer(undefined, { type: DEALCARDS_INVALID });
      expect(result).toMatchSnapshot();
    });
  });

  describe('DEAL_HAND', () => {
    const state = {
      ...cloneDeep(initialState),
      stock: [
        'hearts-A',
        'hearts-2',
        'hearts-3',
        'hearts-4',
        'hearts-5',
        'hearts-6',
        'hearts-7',
        'hearts-8',
      ],
    };

    const action = { type: DEAL_HAND };
    test('should remove 3 cards from the front of the stock array', () => {
      const { stock } = reducer(state, action);
      expect(stock).toMatchSnapshot();
    });
    
    test('should add 3 cards to the beginning of the hand array', () => {
      const { hand } = reducer(state, action);
      expect(hand).toMatchSnapshot();
    })

    test('should not do anything if stock is empty', () => {
      const mockState: State = {
        ...cloneDeep(initialState),
        stock: [],
        hand: ['hearts-A']
      };

      const { hand, stock } = reducer(mockState, action)
      expect(stock).toEqual([])
      expect(hand).toEqual(['hearts-A'])
    }) 
  });
});
