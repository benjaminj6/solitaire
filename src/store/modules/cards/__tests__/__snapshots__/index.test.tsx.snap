// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`reducer DEAL_HAND should add 3 cards to the beginning of the hand array 1`] = `
Array [
  "hearts-A",
  "hearts-2",
  "hearts-3",
]
`;

exports[`reducer DEAL_HAND should remove 3 cards from the front of the stock array 1`] = `
Array [
  "hearts-4",
  "hearts-5",
  "hearts-6",
  "hearts-7",
  "hearts-8",
]
`;

exports[`reducer DEALERCARDS_INVALID should return the state 1`] = `
Object {
  "dealt": false,
  "deck": Object {
    "clubs-10": Object {
      "id": "clubs-10",
      "rank": "10",
      "suit": "clubs",
      "value": 10,
      "visible": false,
    },
    "clubs-2": Object {
      "id": "clubs-2",
      "rank": "2",
      "suit": "clubs",
      "value": 2,
      "visible": false,
    },
    "clubs-3": Object {
      "id": "clubs-3",
      "rank": "3",
      "suit": "clubs",
      "value": 3,
      "visible": false,
    },
    "clubs-4": Object {
      "id": "clubs-4",
      "rank": "4",
      "suit": "clubs",
      "value": 4,
      "visible": false,
    },
    "clubs-5": Object {
      "id": "clubs-5",
      "rank": "5",
      "suit": "clubs",
      "value": 5,
      "visible": false,
    },
    "clubs-6": Object {
      "id": "clubs-6",
      "rank": "6",
      "suit": "clubs",
      "value": 6,
      "visible": false,
    },
    "clubs-7": Object {
      "id": "clubs-7",
      "rank": "7",
      "suit": "clubs",
      "value": 7,
      "visible": false,
    },
    "clubs-8": Object {
      "id": "clubs-8",
      "rank": "8",
      "suit": "clubs",
      "value": 8,
      "visible": false,
    },
    "clubs-9": Object {
      "id": "clubs-9",
      "rank": "9",
      "suit": "clubs",
      "value": 9,
      "visible": false,
    },
    "clubs-A": Object {
      "id": "clubs-A",
      "rank": "A",
      "suit": "clubs",
      "value": 1,
      "visible": false,
    },
    "clubs-J": Object {
      "id": "clubs-J",
      "rank": "J",
      "suit": "clubs",
      "value": 11,
      "visible": false,
    },
    "clubs-K": Object {
      "id": "clubs-K",
      "rank": "K",
      "suit": "clubs",
      "value": 13,
      "visible": false,
    },
    "clubs-Q": Object {
      "id": "clubs-Q",
      "rank": "Q",
      "suit": "clubs",
      "value": 12,
      "visible": false,
    },
    "diamonds-10": Object {
      "id": "diamonds-10",
      "rank": "10",
      "suit": "diamonds",
      "value": 10,
      "visible": false,
    },
    "diamonds-2": Object {
      "id": "diamonds-2",
      "rank": "2",
      "suit": "diamonds",
      "value": 2,
      "visible": false,
    },
    "diamonds-3": Object {
      "id": "diamonds-3",
      "rank": "3",
      "suit": "diamonds",
      "value": 3,
      "visible": false,
    },
    "diamonds-4": Object {
      "id": "diamonds-4",
      "rank": "4",
      "suit": "diamonds",
      "value": 4,
      "visible": false,
    },
    "diamonds-5": Object {
      "id": "diamonds-5",
      "rank": "5",
      "suit": "diamonds",
      "value": 5,
      "visible": false,
    },
    "diamonds-6": Object {
      "id": "diamonds-6",
      "rank": "6",
      "suit": "diamonds",
      "value": 6,
      "visible": false,
    },
    "diamonds-7": Object {
      "id": "diamonds-7",
      "rank": "7",
      "suit": "diamonds",
      "value": 7,
      "visible": false,
    },
    "diamonds-8": Object {
      "id": "diamonds-8",
      "rank": "8",
      "suit": "diamonds",
      "value": 8,
      "visible": false,
    },
    "diamonds-9": Object {
      "id": "diamonds-9",
      "rank": "9",
      "suit": "diamonds",
      "value": 9,
      "visible": false,
    },
    "diamonds-A": Object {
      "id": "diamonds-A",
      "rank": "A",
      "suit": "diamonds",
      "value": 1,
      "visible": false,
    },
    "diamonds-J": Object {
      "id": "diamonds-J",
      "rank": "J",
      "suit": "diamonds",
      "value": 11,
      "visible": false,
    },
    "diamonds-K": Object {
      "id": "diamonds-K",
      "rank": "K",
      "suit": "diamonds",
      "value": 13,
      "visible": false,
    },
    "diamonds-Q": Object {
      "id": "diamonds-Q",
      "rank": "Q",
      "suit": "diamonds",
      "value": 12,
      "visible": false,
    },
    "hearts-10": Object {
      "id": "hearts-10",
      "rank": "10",
      "suit": "hearts",
      "value": 10,
      "visible": false,
    },
    "hearts-2": Object {
      "id": "hearts-2",
      "rank": "2",
      "suit": "hearts",
      "value": 2,
      "visible": false,
    },
    "hearts-3": Object {
      "id": "hearts-3",
      "rank": "3",
      "suit": "hearts",
      "value": 3,
      "visible": false,
    },
    "hearts-4": Object {
      "id": "hearts-4",
      "rank": "4",
      "suit": "hearts",
      "value": 4,
      "visible": false,
    },
    "hearts-5": Object {
      "id": "hearts-5",
      "rank": "5",
      "suit": "hearts",
      "value": 5,
      "visible": false,
    },
    "hearts-6": Object {
      "id": "hearts-6",
      "rank": "6",
      "suit": "hearts",
      "value": 6,
      "visible": false,
    },
    "hearts-7": Object {
      "id": "hearts-7",
      "rank": "7",
      "suit": "hearts",
      "value": 7,
      "visible": false,
    },
    "hearts-8": Object {
      "id": "hearts-8",
      "rank": "8",
      "suit": "hearts",
      "value": 8,
      "visible": false,
    },
    "hearts-9": Object {
      "id": "hearts-9",
      "rank": "9",
      "suit": "hearts",
      "value": 9,
      "visible": false,
    },
    "hearts-A": Object {
      "id": "hearts-A",
      "rank": "A",
      "suit": "hearts",
      "value": 1,
      "visible": false,
    },
    "hearts-J": Object {
      "id": "hearts-J",
      "rank": "J",
      "suit": "hearts",
      "value": 11,
      "visible": false,
    },
    "hearts-K": Object {
      "id": "hearts-K",
      "rank": "K",
      "suit": "hearts",
      "value": 13,
      "visible": false,
    },
    "hearts-Q": Object {
      "id": "hearts-Q",
      "rank": "Q",
      "suit": "hearts",
      "value": 12,
      "visible": false,
    },
    "spades-10": Object {
      "id": "spades-10",
      "rank": "10",
      "suit": "spades",
      "value": 10,
      "visible": false,
    },
    "spades-2": Object {
      "id": "spades-2",
      "rank": "2",
      "suit": "spades",
      "value": 2,
      "visible": false,
    },
    "spades-3": Object {
      "id": "spades-3",
      "rank": "3",
      "suit": "spades",
      "value": 3,
      "visible": false,
    },
    "spades-4": Object {
      "id": "spades-4",
      "rank": "4",
      "suit": "spades",
      "value": 4,
      "visible": false,
    },
    "spades-5": Object {
      "id": "spades-5",
      "rank": "5",
      "suit": "spades",
      "value": 5,
      "visible": false,
    },
    "spades-6": Object {
      "id": "spades-6",
      "rank": "6",
      "suit": "spades",
      "value": 6,
      "visible": false,
    },
    "spades-7": Object {
      "id": "spades-7",
      "rank": "7",
      "suit": "spades",
      "value": 7,
      "visible": false,
    },
    "spades-8": Object {
      "id": "spades-8",
      "rank": "8",
      "suit": "spades",
      "value": 8,
      "visible": false,
    },
    "spades-9": Object {
      "id": "spades-9",
      "rank": "9",
      "suit": "spades",
      "value": 9,
      "visible": false,
    },
    "spades-A": Object {
      "id": "spades-A",
      "rank": "A",
      "suit": "spades",
      "value": 1,
      "visible": false,
    },
    "spades-J": Object {
      "id": "spades-J",
      "rank": "J",
      "suit": "spades",
      "value": 11,
      "visible": false,
    },
    "spades-K": Object {
      "id": "spades-K",
      "rank": "K",
      "suit": "spades",
      "value": 13,
      "visible": false,
    },
    "spades-Q": Object {
      "id": "spades-Q",
      "rank": "Q",
      "suit": "spades",
      "value": 12,
      "visible": false,
    },
  },
  "foundation": Object {
    "clubs": Array [],
    "diamonds": Array [],
    "hearts": Array [],
    "spades": Array [],
  },
  "hand": Array [],
  "stock": Array [],
  "tableau": Object {
    "1": Array [],
    "2": Array [],
    "3": Array [],
    "4": Array [],
    "5": Array [],
    "6": Array [],
    "7": Array [],
  },
}
`;
